arch/risc-v/src/espressif/esp_tickless.c

```diff
@@ -0,0 +1,432 @@
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <assert.h>
+#include <stdint.h>
+#include <time.h>
+
+#include <arch/board/board.h>
+#include <arch/irq.h>
+#include <nuttx/arch.h>
+#include <nuttx/clock.h>
+
+#include "chip.h"
+#include "esp_irq.h"
+
+#include "esp_attr.h"
+#include "hal/systimer_hal.h"
+#include "hal/systimer_ll.h"
+#include "periph_ctrl.h"
+#include "systimer.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#if SOC_SYSTIMER_INT_LEVEL
+#  define SYSTIMER_TRIGGER_TYPE ESP_IRQ_TRIGGER_LEVEL
+#else
+#  define SYSTIMER_TRIGGER_TYPE ESP_IRQ_TRIGGER_EDGE
+#endif /* SOC_SYSTIMER_INT_LEVEL */
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* Systimer HAL layer object */
+
+static systimer_hal_context_t systimer_hal;
+
+/* Whether an interval timer is being started */
+
+static bool g_timer_started;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp_tickless_isr
+ *
+ * Description:
+ *   Handler to be executed by the Systimer ISR.
+ *
+ * Input Parameters:
+ *   irq           - IRQ associated to that interrupt.
+ *   context       - Interrupt register state save info.
+ *   arg           - A pointer to the argument provided when the interrupt
+ *                   was registered.
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+
+static int esp_tickless_isr(int irq, void *context, void *arg)
+{
+  g_timer_started = false;
+
+  systimer_ll_clear_alarm_int(systimer_hal.dev,
+                              SYSTIMER_ALARM_OS_TICK_CORE0);
+
+  nxsched_timer_expiration();
+
+  return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_get_idletime
+ *
+ * Description:
+ *   This function returns the idle time.
+ *
+ * Input Parameters:
+ *   None.
+ *
+ * Returned Value:
+ *   The time in system ticks remaining for idle.
+ *   Zero means system is busy.
+ *
+ ****************************************************************************/
+
+uint32_t up_get_idletime(void)
+{
+  uint32_t us;
+  uint64_t alarm_value;
+  uint64_t counter;
+  irqstate_t flags;
+
+  flags = enter_critical_section();
```

Comment:
should this be a critical section or a spinlock?

Reply:
I initially opted for the "safest" approach,  I used a critical section because it has been used in other chips for Tickless implementation.

But I believe the `spin_lock_irqsave/spin_unlock_irqrestore` API fits here, since the critical section intends to ensure the atomicity of the time retrieval operation and to protect the SYSTIMER registers from concurrent accesses.